### DISCUSSION TITLE
Is there a way to convert Data Type in a FA?

### DISCUSSION THREAD

**Discussion by CRohm42** [fluentassertions/fluentassertions#2232] [Q&A]
I'm guessing there isn't a way to convert data types in the assertion, but wanted to make sure we're not missing a better way to perform an assertion for a specific scenario.

We commonly perform grid column sort order tests with List.Should().BeInDescending/AscendingOrder().  The grid column values are always strings (grabbing off GUI), so the process is as follows:

-Sort column
-Logic to determine column data type
-Create list, converting from string to "target" data type of int/currency/date/etc. when needed (if not converted, FA's fail/pass unexpectedly)
-Perform FA

Is there a way for us to simply create the list of strings and then perform the FA and give it the "target" data type so the string list is evaluated as an int/currency/date/etc?

**Comment by jnyrup** [comment]
If you create passing and failing code examples, it would be easier to help you find `Foo`.

```cs
// passing example
var subject = new ...
var expected = new ...

subject.Should().Foo(expected);
```

```
// failing example
var subject = new ...
var unexpected = new ...

subject.Should().Foo(unexpected);
```

**Reply by CRohm42** [reply]
````cs
// Grid column values are strings, no conversion needed when creating list from grid
List<string> columnDataAsStrings = new List<string>() { "657147", "65803", "65807" };  

// Must convert string values to integer when creating list from grid
List<int> columnDataAsInts= new List<int>() { 657147, 65803, 65807 };  

columnDataAsStrings.Should().BeInAscendingOrder();  // Passes unexpectedly
columnDataAsInts.Should().BeInAscendingOrder();  // Fails as expected
````

Is there a way to change:
`columnDataAsStrings.Should().BeInAscendingOrder();`

so that it evaluates like:
`columnDataAsInts.Should().BeInAscendingOrder();`

To prevent the data conversion prior to FA.

Hopefully this helps!

**Reply by jnyrup** [reply]
You can provide a custom comparer to `BeInAscendingOrder` to tell it how to compare two `string`s.
Here's the simply example that just assumes that all items in `columnDataAsStrings` should be interpret as `int`s.
You can extend the body of `Compare(string? x, string? y)` to handle when items in `columnDataAsStrings` can be of different types.

```cs
var columnDataAsStrings = new List<string>() { "657147", "65803", "65807" };
columnDataAsStrings.Should().BeInAscendingOrder(new MyComparer());

sealed class MyComparer : IComparer<string>
{
    public int Compare(string? x, string? y) => int.Parse(x).CompareTo(int.Parse(y));
}
```

**Comment by IT-VBFK** [comment]
Something like:
https://fluentassertions.com/objectgraphs/#equivalency-comparison-behavior

or even
https://fluentassertions.com/objectgraphs/#auto-conversion ?
